pkg/cloud/services/elb/loadbalancer.go

```diff
@@ -54,6 +55,340 @@ const maxELBsDescribeTagsRequest = 20
+			return errors.Wrapf(err, "failed to reconcile tags for apiserver load balancer %q", lb.Name)
+		}
+
+		// Reconcile the subnets and availability zones from the spec
+		// and the ones currently attached to the load balancer.
+		if len(lb.SubnetIDs) != len(spec.SubnetIDs) {
+			_, err := s.ELBV2Client.SetSubnets(&elbv2.SetSubnetsInput{
+				LoadBalancerArn: &lb.ARN,
+				Subnets:         aws.StringSlice(spec.SubnetIDs),
+			})
+			if err != nil {
+				return errors.Wrapf(err, "failed to set subnets for apiserver load balancer '%s'", lb.Name)
+			}
+		}
+		if len(lb.AvailabilityZones) != len(spec.AvailabilityZones) {
+			lb.AvailabilityZones = spec.AvailabilityZones
+		}
+
+		// Reconcile the security groups from the spec and the ones currently attached to the load balancer
+		if s.scope.ControlPlaneLoadBalancer().LoadBalancerType != infrav1.LoadBalancerTypeNLB && !sets.NewString(lb.SecurityGroupIDs...).Equal(sets.NewString(spec.SecurityGroupIDs...)) {
+			_, err := s.ELBV2Client.SetSecurityGroups(&elbv2.SetSecurityGroupsInput{
+				LoadBalancerArn: &lb.ARN,
+				SecurityGroups:  aws.StringSlice(spec.SecurityGroupIDs),
+			})
+			if err != nil {
+				return errors.Wrapf(err, "failed to apply security groups to load balancer %q", lb.Name)
+			}
+		}
+	} else {
+		s.scope.Trace("Unmanaged control plane load balancer, skipping load balancer configuration", "api-server-elb", lb)
+	}
+	lb.DeepCopyInto(&s.scope.Network().APIServerELB)
+	return nil
+}
+
+func (s *Service) getAPIServerLBSpec(elbName string) (*infrav1.LoadBalancer, error) {
+	var securityGroupIDs []string
+	controlPlaneLoadBalancer := s.scope.ControlPlaneLoadBalancer()
+	if controlPlaneLoadBalancer != nil && controlPlaneLoadBalancer.LoadBalancerType != infrav1.LoadBalancerTypeNLB {
+		securityGroupIDs = append(securityGroupIDs, controlPlaneLoadBalancer.AdditionalSecurityGroups...)
+		securityGroupIDs = append(securityGroupIDs, s.scope.SecurityGroups()[infrav1.SecurityGroupAPIServerLB].ID)
+	}
+
+	res := &infrav1.LoadBalancer{
+		Name:         elbName,
+		Scheme:       s.scope.ControlPlaneLoadBalancerScheme(),
+		V2Attributes: make(map[string]*string),
+		V2Listeners: []infrav1.Listener{
+			{
+				Protocol: infrav1.ElbProtocolTCP,
+				Port:     6443,
+				TargetGroup: infrav1.TargetGroupSpec{
+					Name:     aws.String(fmt.Sprintf("apiserver-target-%d", time.Now().Unix())),
+					Port:     aws.Int64(6443),
+					Protocol: infrav1.ElbProtocolTCP,
+					VpcID:    aws.String(s.scope.VPC().ID),
+					HealthCheck: &infrav1.TargetGroupHealthCheck{
+						HealthCheckProtocol: aws.String(string(infrav1.ElbProtocolTCP)),
+						HealthCheckPort:     aws.String("6443"),
+					},
+				},
+			},
+		},
+		SecurityGroupIDs: securityGroupIDs,
+	}
+
+	if s.scope.ControlPlaneLoadBalancer() != nil && s.scope.ControlPlaneLoadBalancer().LoadBalancerType != infrav1.LoadBalancerTypeNLB {
+		res.V2Attributes[infrav1.LoadBalancerAttributeIdleTimeTimeoutSeconds] = aws.String("600")
```

Comment:
nit, could the 600 be a constant? Also where does the value come from, is it from AWS?

Reply:
Oh good catch! Also, that should have been `60` and not `600`. :D Indeed, it comes from the docs here: https://docs.aws.amazon.com/elasticloadbalancing/latest/APIReference/API_LoadBalancerAttribute.html